pyro/distributions/von_mises.py

```diff
@@ -64,13 +104,43 @@ def __init__(self, loc, concentration, validate_args=None):
         self.loc, self.concentration = broadcast_all(loc, concentration)
         batch_shape = self.loc.shape
         event_shape = torch.Size()
+
+        # Moments
+        self._variance = 1 - (_log_modified_bessel_fn(self.concentration, order=1) -
+                              _log_modified_bessel_fn(self.concentration, order=0)).exp()
+
+        # Parameters for sampling
+        tau = 1 + (1 + 4 * self.concentration ** 2).sqrt()
+        rho = (tau - (2 * tau).sqrt()) / (2 * self.concentration)
+        self._proposal_r = (1 + rho ** 2) / (2 * rho)
+
         super(VonMises, self).__init__(batch_shape, event_shape, validate_args)
 
     def log_prob(self, value):
         log_prob = self.concentration * torch.cos(value - self.loc)
-        log_prob = log_prob - math.log(2 * math.pi) - _log_modified_bessel_fn_0(self.concentration)
+        log_prob = log_prob - math.log(2 * math.pi) - _log_modified_bessel_fn(self.concentration, order=0)
         return log_prob
 
+    def sample(self, sample_shape=torch.Size()):
```

Comment:
It seems like we should also be able to differentiate through this. Correct me if I am wrong. If so, we should make this `rsample`, and add `has_rsample = True`.

Reply:
Ah, I understood it as `rsample` being completely reparametrizable (instead of just differentiable). The Von Mises distribution is not reparametrizable w.r.t. concentration which is why I picked `sample`, but I will change it to rsample now.

Reply:
Sorry, you are right - this is more involved than the accept-reject sampler in algorithm 1 from https://arxiv.org/pdf/1610.05683.pdf.